### PR TITLE
Apache http client fix exception handling

### PR DESCRIPTION
**Pull request by mar-kolya**
This should improve how 'hard' cases like redirects and errors are handled by apache http client instrumentation.

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * An element matcher that matches a super type. Exceptions during matching process are logged and
+   * ignored.
+   *
+   * @param <T> The type of the matched entity.
+   * @see net.bytebuddy.matcher.HasSuperTypeMatcher
+   */
+  @HashCodeAndEqualsPlugin.Enhance
+  public static class SafeHasSuperTypeMatcher<T extends TypeDescription>
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java]: Can you include in your comment how this is different from the `SuperTypeMatcher` that is built into byte buddy?

Also describe how this is different from using `failSafe(hasSuperType(...))`.

Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java]: should be fixed now

Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java]: much better.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+   *
+   * @param matcher The type to be checked for being a super type of the matched type.
+   * @param <T> The type of the matched object.
+   * @return A matcher that matches any type description that declares a super type that matches the
+   *     provided matcher.
+   * @see ElementMatchers#hasSuperType(net.bytebuddy.matcher.ElementMatcher)
+   */
+  public static <T extends TypeDescription> ElementMatcher.Junction<T> safeHasSuperType(
+      final ElementMatcher<? super TypeDescription> matcher) {
+    // return hasSuperType(matcher);
```
[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ByteBuddyElementMatchers.java]: Good catch! Removed.

**Pull request merged**